### PR TITLE
Ericjohnson327/xmeta tests

### PR DESCRIPTION
**Pull request by EricJohnson327**

### CODE REVIEW

File: src/test/xmeta/test_xmeta_emit.cpp [hunk 1]
```diff
@@ -283,16 +283,403 @@ std::vector<uint8_t> run_and_save_to_memory(std::istringstream & test_idl, std::
...
+        VerifyTypeRefName(expectedTypeRef, parent.TypeNamespace(), std::get<coded_index<TypeDefOrRef>>(methodSig.ReturnType().Type().Type()).TypeRef());
+        break;
+    }
+    default:
+        REQUIRE(std::get<ElementType>(methodSig.ReturnType().Type().Type()) == expectedType);
+        break;
+    }
+}
+
+struct ExpectedType {
```
Comment (kythant) [inline src/test/xmeta/test_xmeta_emit.cpp]: Open bracket should be on new line. Same for the rest.

File: src/test/xmeta/test_xmeta_emit.cpp [hunk 2]
```diff
@@ -283,16 +283,403 @@ std::vector<uint8_t> run_and_save_to_memory(std::istringstream & test_idl, std::
...
+            if (method.Flags().SpecialName() == true)
+            {
+                if (getMethodName == method.Name())
+                {
+                    foundGetMethod = true;
+                }
+                else if (setMethodName == method.Name())
+                {
+                    foundSetMethod = true;
+                    REQUIRE(!method.Signature().ReturnType());
```
Comment (kythant) [inline src/test/xmeta/test_xmeta_emit.cpp]: We should check the signature of the formal parameter for the set method here.

File: src/test/xmeta/test_xmeta_emit.cpp [hunk 3]
```diff
@@ -283,16 +283,403 @@ std::vector<uint8_t> run_and_save_to_memory(std::istringstream & test_idl, std::
...
+        std::string setMethodName = "put_";
+        getMethodName += prop.Name();
+        setMethodName += prop.Name();
+        for (auto &&method : prop.Parent().MethodList())
+        {
+            if (method.Flags().SpecialName() == true)
+            {
+                if (getMethodName == method.Name())
+                {
+                    foundGetMethod = true;
```
Comment (kythant) [inline src/test/xmeta/test_xmeta_emit.cpp]: We should check the signature of the return type here for the get method. It should be the type of the property.

File: src/test/xmeta/test_xmeta_emit.cpp [hunk 4]
```diff
@@ -283,16 +283,403 @@ std::vector<uint8_t> run_and_save_to_memory(std::istringstream & test_idl, std::
...
+            if (method.Flags().SpecialName() == true)
+            {
+                if (addMethodName == method.Name())
+                {
+                    foundAddMethod = true;
+                    REQUIRE(size(method.Signature().Params()) == 1);
+
+                    std::string expectedTypeName(std::get<coded_index<TypeDefOrRef>>(method.Signature().Params().first[0].Type().Type()).TypeRef().TypeName());
+                    VerifyTypeRefName(expectedTypeName, event.Parent().TypeNamespace(), event.EventType().TypeRef());
+                    REQUIRE(method.Flags().value == interface_method_event_attributes().value);
```
Comment (kythant) [inline src/test/xmeta/test_xmeta_emit.cpp]: Need to check the signature of the formal parameter. Particularly the add method has a return sig with type ref EventRegistrationToken and the remove method has formal parameter signature with type ref EventRegistrationToken. I don't see anything referring and checking for EventRegistrationToken. 

Look at this for example

    REQUIRE(add_method.Name() == "add_Changed");	
    {	
        auto const& sig = add_method.Signature();	
        REQUIRE(std::get<coded_index<TypeDefOrRef>>(sig.ReturnType().Type().Type()).TypeRef() == event_registration_token_ref);	
        REQUIRE(size(sig.Params()) == 1);	
        auto const& param_sig = sig.Params().first[0];	
        REQUIRE(std::get<coded_index<TypeDefOrRef>>(param_sig.Type().Type()).TypeRef() == string_list_event_ref);	
        REQUIRE(add_method.Flags().value == interface_method_event_attributes().value);	
    }	
    auto const& remove_method = event1.MethodSemantic().first[1].Method();	
    REQUIRE(remove_method.Name() == "remove_Changed");	
    {	    {
        auto const& sig = remove_method.Signature();	        std::make_unique<ExpectedInterface>("IComboBox",
        REQUIRE(!sig.ReturnType());	            std::vector<ExpectedEvent> {
        REQUIRE(size(sig.Params()) == 1);	                ExpectedEvent("Changed", "StringListEvent")
        auto const& param_sig = sig.Params().first[0];	            }
        REQUIRE(std::get<coded_index<TypeDefOrRef>>(param_sig.Type().Type()).TypeRef() == event_registration_token_ref);	        ),
        REQUIRE(remove_method.Flags().value == interface_method_event_attributes().value);	        std::make_unique<ExpectedDelegate>("StringListEvent", ElementType::Void,
    }

File: src/test/xmeta/test_xmeta_emit.cpp [hunk 5]
```diff
@@ -283,16 +283,403 @@ std::vector<uint8_t> run_and_save_to_memory(std::istringstream & test_idl, std::
...
-    REQUIRE(ref.TypeNamespace() == namespace_name);
-    return ref;
+        combineName += typeRef.TypeNamespace();
+        combineName += ".";
+    }
+    combineName += typeRef.TypeName();
+    REQUIRE(combineName == expectedName);
+}
+
+void VerifyReturnType(const ElementType &expectedType, const std::string &expectedTypeRef, const MethodDef &method, const TypeDef &parent)
```
Comment (kennykerr) [inline src/test/xmeta/test_xmeta_emit.cpp]: Place `const` consistently: `ElementType const& expectedType`

File: src/test/xmeta/test_xmeta_emit.cpp [hunk 6]
```diff
@@ -283,16 +283,403 @@ std::vector<uint8_t> run_and_save_to_memory(std::istringstream & test_idl, std::
     return reader.save_to_memory();
 }
 
-template<typename T>
-auto find_type_by_name(table<T> const& type, std::string name, std::string namespace_name)
+void VerifyTypeRefName(std::string expectedName, std::string_view currentNamespace, TypeRef const& typeRef)
```
Comment (kennykerr) [inline src/test/xmeta/test_xmeta_emit.cpp]: `const&` expectedName or `std::string_view`

File: src/test/xmeta/test_xmeta_emit.cpp [hunk 7]
```diff
@@ -283,16 +283,403 @@ std::vector<uint8_t> run_and_save_to_memory(std::istringstream & test_idl, std::
...
+}
+
+void VerifyReturnType(const ElementType &expectedType, const std::string &expectedTypeRef, const MethodDef &method, const TypeDef &parent)
+{
+    auto const& methodSig = method.Signature();
+    REQUIRE(method.ParamList().first[0].Name() == "returnVal");
+    switch (expectedType)
+    {
+    case ElementType::Void:
+
```
Comment (kennykerr) [inline src/test/xmeta/test_xmeta_emit.cpp]: Weird spacing.

File: src/test/xmeta/test_xmeta_emit.cpp [hunk 8]
```diff
@@ -283,16 +283,403 @@ std::vector<uint8_t> run_and_save_to_memory(std::istringstream & test_idl, std::
...
+        {
+            REQUIRE(field.Signature().Type().element_type() == type);
+        }
+    }
+};
+
+struct ExpectedStruct : ExpectedType {
+    std::vector<ExpectedStructField> fields;
+
+    ExpectedStruct(const std::string &expectedName, std::vector<ExpectedStructField> &&expectedFields) : ExpectedType(expectedName), fields(expectedFields) {}
```
Comment (kennykerr) [inline src/test/xmeta/test_xmeta_emit.cpp]: Why std::vector&& when you are still copying?

File: src/test/xmeta/test_xmeta_emit.cpp [hunk 9]
```diff
@@ -283,16 +283,427 @@ std::vector<uint8_t> run_and_save_to_memory(std::istringstream & test_idl, std::
...
+void VerifyReturnType(ElementType const& expectedType, std::string const& expectedTypeRef, MethodDef const& method, TypeDef const& parent)
+{
+    auto const& methodSig = method.Signature();
+    REQUIRE(method.ParamList().first[0].Name() == "returnVal");
+    switch (expectedType)
+    {
+    case ElementType::Void:
+        REQUIRE(!methodSig.ReturnType());
+        break;
+    case ElementType::End:
```
Comment (DefaultRyan) [inline src/test/xmeta/test_xmeta_emit.cpp]: This method seems a bit clunky. Why is ElementType::End involved? Why is expectedTypeRef only checked in the ElementType::End case? I fear this won't really scale, especially when we start handling generic types.

Comment (EricJohnson327) [reply src/test/xmeta/test_xmeta_emit.cpp]: Will fix later.  This works with currently supported types.  Opened a bug to refactor this to more easily support generics/nested types in the future.

**Pull request merged**